### PR TITLE
refactor: use glmnet instead of sklearn for path

### PR DESCRIPTION
**Pull request by jolars**
Right now we're using manual path-fitting code to first find a reasonable path, which is based on the same criteria that glmnet uses, but instead we use sklearn. The problem is that sklearn is much slower than glmnet for many problems (due to screening) which currently slows down everything.

This PR switches to use glmnet instead. The downside is that it adds more requirements to the benchmark since r and glmnet is always required, but it speeds everything up and removes some code that we no longer have to maintain.

### CODE REVIEW

File: objective.py [hunk 1]
```diff
@@ -29,63 +38,30 @@ def set_data(self, X, y):
...
+
+        y_tmp = robjects.FloatVector(y)
+        glmnet = robjects.r["glmnet"]
+        glmnet_fit = glmnet(
+            X_tmp,
+            y_tmp,
+            intercept=self.fit_intercept,
+            standardize=False,
+            maxit=1_000_000,
         )
```
Comment (tomMoral) [inline objective.py]: don't you have the value of the results also with this call to `glmnet`? it would be nice to add the computation of the iterates convergence with these?

I would advise to take this computation in a standalone function, and use a caching mechanism to call this? We can discuss about this during the sprint if you want.

**Comment by jolars** [comment]
> don't you have the value of the results also with this call to `glmnet`? it would be nice to add the computation of the iterates convergence with these?
That's right, but what do you mean by "add the computation of the iterates convergence"?

> I would advise to take this computation in a standalone function, and use a caching mechanism to call this? We can discuss about this during the sprint if you want.

Sure, that sounds like a good idea! But of course it's slightly complicated because the result differs depending on both the dataset parameters and the objective parameters. This is what I was referring to in https://github.com/benchopt/benchopt/issues/454.

**Comment by tomMoral** [comment]
> "add the computation of the iterates convergence"

I mean compute `|x(lmbd) - x*(lmbd)|` for instance in the loss, and also the recovery of the support?

For the caching, adding and calling such independent function would solve the issue no?

```python
@cache
def get_glmnet_best_result(X, y, fit_intercept):
    # run glmnet once to obtain a lambda sequence
    if sparse.issparse(X):
            r_Matrix = packages.importr("Matrix")
            X_coo = X.tocoo()
            X_tmp = r_Matrix.sparseMatrix(
                i=robjects.IntVector(X_coo.row + 1),
                j=robjects.IntVector(X_coo.col + 1),
                x=robjects.FloatVector(X_coo.data),
                dims=robjects.IntVector(X_coo.shape),
            )
    else:
            X_tmp = robjects.r.matrix(X, X.shape[0], X.shape[1])

    y_tmp = robjects.FloatVector(y)
    glmnet = robjects.r["glmnet"]
    glmnet_fit = glmnet(
            X_tmp,
            y_tmp,
            intercept=self.fit_intercept,
            standardize=False,
            maxit=1_000_000,
     )
    return dict(zip(glmnet_fit.names, list(glmnet_fit)))
```

**Comment by jolars** [comment]
> I mean compute `|x(lmbd) - x*(lmbd)|` for instance in the loss, and also the recovery of the support?

Sure, but I'm not sure we can treat this as a gold standard unless we set the tolerance very low, which might inflate the run time.

> For the caching, adding and calling such independent function would solve the issue no?

I didn't think it would be this simple! So you can actually cache on the entire dataset. It seems like that would make for very expensive caching, but I suppose maybe you mean that instead of X and y we use the name of the data set?

**Comment by tomMoral** [comment]
Oh, it seemed to me that this was the goal here was running `glmnet` with a low tolerance. If this is not the case, nevermind :)

For caching, this is not so expansive as this only requires hashing `X, y` once per run and numpy object hash is relatively reasonable. So I would say this is a good temporary fix.
But indeed, finding a better solution were we hash the `dataset` (which is cheap to hash) would make sense, but we can discuss this in [benchopt/benchopt#454](https://github.com/benchopt/benchopt/issues/454).

**Comment by jolars** [comment]
> Oh, it seemed to me that this was the goal here was running `glmnet` with a low tolerance. If this is not the case, nevermind :)

No the goal is actually just to find a sequence of reasonable lambda values by using early-stopping criteria for the path.

> For caching, this is not so expansive as this only requires hashing `X, y` once per run and numpy object hash is relatively reasonable. So I would say this is a good temporary fix.

Okay, great, thanks! I guess this applies to scipy sparse matrices too? (Maybe scipy uses numpy structures under the hood?)

**Comment by mathurinm** [comment]
Yes scipy stores data values, indices and indices pointers (for CSC and CSR) in numpy arrays

**Comment by jolars** [comment]
Actually, it's not possible to have either numpy or scipy arguments when caching, since neither are hashable. 

I tried using cachetools instead, for which you can specify which arguments you cache on, but then it would fail to cache altogether. In the end, glmnet is fast enough at this tolerance to not warrant caching at all, so I'm just skipping it altogether.
